language/tools/move-cli/src/main.rs

```diff
@@ -391,14 +406,39 @@ fn run(
     } else {
         let effects = session.finish().map_err(|e| e.into_vm_status())?;
         if verbose {
-            explain_effects(&effects, &state)?
+            explain_execution_effects(&effects, &state)?
         }
-        maybe_commit_effects(!dry_run, Some(effects), &state)
+        maybe_commit_effects(!dry_run, effects, &state)
     }
 }
 
-fn explain_effects(effects: &TransactionEffects, state: &OnDiskStateView) -> Result<()> {
-    // all module publishing happens via save_modules(), so effects shouldn't contain modules
+fn get_cost_strategy(gas_budget: Option<u64>) -> Result<CostStrategy<'static>> {
+    let gas_schedule = &vm_genesis::genesis_gas_schedule::INITIAL_GAS_SCHEDULE;
+    let cost_strategy = if let Some(gas_budget) = gas_budget {
+        let max_gas_budget = u64::MAX
+            .checked_div(gas_schedule.gas_constants.gas_unit_scaling_factor)
+            .unwrap();
+        if gas_budget >= max_gas_budget {
+            bail!("Gas budget set too high; maximum is {}", max_gas_budget)
+        }
+        CostStrategy::transaction(gas_schedule, GasUnits::new(gas_budget))
+    } else {
+        // no budget specified. use CostStrategy::system, which disables gas metering
+        CostStrategy::system(gas_schedule, GasUnits::new(0))
+    };
+    Ok(cost_strategy)
+}
+
+fn explain_publish_effects(effects: &TransactionEffects, _state: &OnDiskStateView) -> Result<()> {
+    // publish effects should contain no events and resources
+    assert!(effects.events.is_empty());
+    assert!(effects.resources.is_empty());
+    // TODO
```

Comment:
Finish explaining what's left to be done + how it will use `state`?

Reply:
Seems that this PR is in the right direction, I'll finish up with the explanation.

Reply:
Done, with the recent commit. For each module in the `TransactionEffects`, it explains whether this is `Publishing a new module <module_id>` or `Updating an existing module <module_id>`.